lib/rubygems/platform.rb

```diff
@@ -171,7 +172,11 @@ def ===(other)
     @os == other.os and
 
     # version
-    (@version.nil? or other.version.nil? or @version == other.version)
+    (
+      (@os != 'linux' and (@version.nil? or other.version.nil?)) or
+      (@os == 'linux' and ((@version.nil? and other.version == 'gnu') or (@version == 'gnu' and other.version == nil))) or
```

Comment:
This code should never happen as of today's platform parser. I included it as a defensive coding practice should the platform parser code evolve into not turning `*-linux-gnu` into `*-linux`.

Reply:
If this is not needed, I'd rather not include it. In any case, aren't you missing the "one of them is nil" case if os is linux? Also you're introducing a small coding style inconsistency be checking for `nil` with `== nil` instead of `nil?`.

Reply:
> In any case, aren't you missing the "one of them is nil" case if os is linux?

Through these cases I did infer that in general `nil` is some sort of wildcard meaning "match anything" (e.g for `darwin` that would be "any darwin-XY version") but in the case of `linux` there's this historically introduced special case that `linux` == `linux-gnu` and nothing else.

That's the source of the overall issue:

- on a musl system such as Alpine (`linux-musl`) it means that current rubygems finds a match in gems marked as `linux` and can attempt to fetch+install them (even if there are `linux-musl` gems published), which is incorrect.
- on a glibc system (`linux` a.k.a `linux-gnu`) it means that current rubygems finds a match in gems marked as `linux-musl` and can attempt to fetch+install them (even if there are `linux` or theoretical `linux-gnu` gems published), which is incorrect.

> Also you're introducing a small coding style inconsistency be checking for nil with == nil instead of nil?

Good catch! Will fix...

Reply:
Resolved by removing the unreachable condition.